experimental/parser/legalize_def.go

```diff
@@ -118,6 +118,45 @@ func legalizeTypeDefLike(p *parser, what taxa.Noun, def ast.DeclDef) {
 // legalizeMessageLike legalizes something that resembles a field definition:
 // namely, fields, groups, and enum values.
 func legalizeFieldLike(p *parser, what taxa.Noun, def ast.DeclDef) {
+	if def.Name().IsZero() {
+		def.MarkCorrupt()
+		p.Errorf("missing name %v", what.In()).Apply(
+			report.Snippet(def),
+		)
+	} else if def.Name().AsIdent().IsZero() {
+		def.MarkCorrupt()
+		p.Error(errUnexpected{
+			what:  def.Name(),
+			where: what.In(),
+			want:  taxa.Ident.AsSet(),
+		})
+	}
+
+	// NOTE: We do not legalize a missing value for fields and enum values
+	// here; instead, that happens during IR lowering. This is because we want
+	// to be able to include a suggested field number, but we cannot do that
+	// until much later, when we have evaluated expressions.
```

Comment:
I know I said that this would be fine, but I think we need a TODO to here to changes things up.

For one, we can't really or safely suggest extension numbers, so should probably diagnose extension fields without numbers unconditionally here. But also, if a caller issues an incremental query for just the AST, it really seems like a missed opportunity that we provide no diagnosis for missing field numbers at this point.

I know we've kicked around the term "deferred diagnosis", but that or something other mechanism to defer parts of it would be much better. I think we'd ideally diagnose the missing field now, but another step (lowering to IR) could find all such diagnoses and then augment them with suggested field numbers later.

Reply:
Here's what I'll do: I'm going to add diagnostics for this without suggestions now, and add a TODO to figure out diagnostic suggestions later.